packages/vite/src/node/plugins/terser.ts

```diff
@@ -1,26 +1,40 @@
-import { dirname } from 'path'
-import { fileURLToPath } from 'url'
 import { Worker } from 'okie'
 import type { Terser } from 'types/terser'
 import type { Plugin } from '../plugin'
 import type { ResolvedConfig } from '..'
+import { requireResolveFromRootWithFallback } from '../utils'
 
-// TODO: use import()
-const _dirname = dirname(fileURLToPath(import.meta.url))
+let terserPath: string | undefined
+const loadTerserPath = (root: string) => {
+  if (terserPath) return terserPath
+  try {
+    terserPath = requireResolveFromRootWithFallback(root, 'terser')
```

Comment:
I think we can use `createRequire` here instead https://nodejs.org/docs/latest-v14.x/api/module.html#module_module_createrequire_filename

Reply:
Actually the one used in the CSS plugin should be changed to `createRequire` too

Reply:
Is it safe to remove resolving from `root`? I now found these related things: https://github.com/vitejs/vite/commit/ddfcbce05753da3e525c1bb9bdf449a322e997f4, https://github.com/vitejs/vite/pull/3988, https://github.com/vitejs/vite/issues/2612.

Reply:
Instead of
```js
const resolved = require.resolve(name, [root, ...fallbackPaths]);
require(resolved)
```

I think it's safer to use:
```js
const requireFromRoot = createRequire(path.resolve(root, 'index.cjs'))
requireFromRoot(name)
```

Reply:
See https://github.com/import-js/eslint-plugin-import/pull/1591

Reply:
Thank you. IIUC it needs to be like this? b6ec807e863e0e27cb16846f995c8e3a5c48b9fd

Reply:
Wait 🤔 This problem is more complicated than I thought…

To clarify, I'm concerned about the `require` method used here is not "clean" (would be interfered with by `require.cache`, etc.).
**Both** `require` and `require.resolve` should be created from the root path so that we can get a fresh instance of `terser`, `sass`, etc. The point is that `require` should be returned from `createRequire(some-file-in-the-root-dir)` too.

Here comes the complexity:
- The above-mentioned refactor will fix a theoretical problem. Though I think it's better, no known issues have been found with the old way.
- But the refactor would break a real-world use case (resolve fallbacks): https://github.com/vitejs/vite/pull/3988
- On the other hand, if I understand correctly, the use case seems to already be broken by the ESM refactor: the `_require` used in `css.ts` now is a newly created function, which doesn't inherit `require.resolve.paths` from the framework that wraps Vite https://github.com/sapphi-red/vite/blob/b6ec807e863e0e27cb16846f995c8e3a5c48b9fd/packages/vite/src/node/plugins/css.ts#L1282

Reply:
Anyway, my original suggestion is to make `requireResolveFromRootWithFallback` `requireFromRootWithFallback` (that is, put the `require` part into the function too).

It's a nice-to-have.

Reply:
I'm leaving this for now because I tried now but I didn't make it work.

Reply:
@sodatea do you think we could merge this then? And then work on your request as an enhancement? Or is this still blocking the PR?

Reply:
Yeah the PR looks good to me. It's not a blocking change.